sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -111,6 +113,23 @@ object DateTimeUtils {
     computedTimeZones.computeIfAbsent(timeZoneId, computeTimeZone)
   }
 
+  private val threadLocalComputedCalendarsMap =
+    new ThreadLocal[mutable.Map[TimeZone, Calendar]] {
+      override def initialValue(): mutable.Map[TimeZone, Calendar] = {
+        mutable.Map[TimeZone, Calendar]()
+      }
+    }
+
+  def getCalendar(timeZone: TimeZone): Calendar = {
+    val c = threadLocalComputedCalendarsMap.get()
+      .getOrElseUpdate(timeZone, {
+        Calendar.getInstance(timeZone)
+      })
+    c.clear()
```

Comment:
Doesn't `clear` reset the timezone of that `Calendar` instance?

Reply:
Nope. It clears all the ```fields``` and zone is not a field.